# Aufgabe 1

## Käsekästchen

#### Zustandsraum
Das 6×4-Raster für Käsekästchen muss zerlegt werden in Kanten zwischen den Feldern und die Felder selbst. Eine Kante kann gezogen oder nicht gezogen sein, ein Feld kann leer oder von einem der beiden Spieler gewonnen sein.

Die Felder F lassen sich wie üblich als zweidimensionale Koordinaten darstellen.
Die Kanten K lassen sich ebenfalls mit zweidimensionalen Koordinaten nach diesem Schema darstellen:
```
 0011223344556
0 - - - - - -
1| | | | | | |
2 - - - - - -
3| | | | | | |
4 - - - - - -
5| | | | | | |
6 - - - - - -
7| | | | | | |
8 - - - - - -
```

Zwischen K und F besteht eine Abhängigkeit: Wenn 1 oder 2 Felder von gezogenen Kanten eingeschlossen sind, bei 2 Feldern ohne Trennung voneinander, so müssen diese Felder genau einem der beiden Spieler zugeordnet sein. Alle anderen Felder müssen leer sein.

#### Aktionsraum
Es gibt eine Aktion genau für jede Kante im Feld.

### Vier gewinnt
#### Zustandsraum
#### Aktionsraum

### Tic Tac Toe
#### Zustandsraum
Das Spiel hat 3×3 Felder, von denen jedes in einem von 3 Zuständen sein kann: leer, besetzt von Spieler 1 und besetzt von Spieler 2. Hierbei muss die Anzahl der von Spieler 1 besetzten Felder gleich oder um 1 größer sein als die Anzahl der von Spieler 2 besetzten Felder.
Weiterhin kann maximal eine Gerade durch das Feld (wenn wir das Spielfeld als F_3^2 betrachten) vollständig von einem Spieler besetzt sein.
#### Aktionsraum
Der Aktionsraum beinhaltet eine Aktion für jedes Feld.

## Vier gewinnt

Wir gehen von einem 7x6 großen Spielfeld aus, welches durch eine Matrix representiert werden kann.

### Zustandsraum
Die Einträge dieser Matrix sind dann 0 (leer), 1 (Spieler 1), 2 (Spieler 2). 

Also ist der Zustandsraum $S$ Teilmenge von $\mathcal{M}_{7\times6}(\mathbb{F}_3)$

Auch hier gibt es wieder einige einschränkende Regeln:
- Es dürfen maximal ein Mal vier ununterbrochene 1en oder 2en in einer Reihe, Spalte oder Diagonale vorkommen, da das Spiel damit endet.
- Mindestens eine dieser Zahlen darf nur 0en darüber haben
- Alle 0en in der Matrix dürfen keine 1en oder 2en darüber haben.

ein Zustand könnte also beispielsweise wie folgt aussehen:

$$
s = \left(\begin{array}{ccccccc}
0 & 0 & 0 & 0 & 0 & 0 & 0 \\
0 & 0 & 0 & 0 & 0 & 0 & 0 \\
0 & 2 & 0 & 0 & 0 & 0 & 0 \\
0 & 1 & 1 & 0 & 0 & 0 & 0 \\
0 & 1 & 2 & 2 & 1 & 0 & 0 \\
2 & 1 & 1 & 1 & 2 & 0 & 0 \\
\end{array}\right)
$$

### Aktionsraum

Alle möglichen Aktionen können durch einen Spaltenindex $\{0, ..., 6\}$ ausgedrückt werden. Der Aktionsraum ist also allgemein $A = \{0, 1, 2, 3, 4, 5, 6\}$.

Während dem Spiel können Aktionen unmöglich werden, wenn die entsprechende Spalte schon voll ist.

## Tic Tac Toe

### Zustandsraum:
es gibt maximal $3^9$ Zustände: jedes der 9 Felder kann 3 Zustände haben: leer, kreis oder kreuz

Diese Zustände können als Tupel mit 9 Einträgen aus $\{0,1,2\}$ dargestellt werden.
Die Menge aller möglichen Zustände ist also definitiv eine Teilmenge von $\mathbb{F}_3 {} ^9$.

- Da das Spiel endet sobald mindestens eine Reihe, Spalte oder Diagonale nur 1en oder 2en enthält, sind viele dieser Zustände nicht möglich (alle mit mehr Folgen gleicher Zahlen ungleich 0)

- Außerdem müssen die Spieler abwechselnd Symbole setzen. Das heißt die Anzahl an 2en und 1en darf sich maximal um 1 unterscheiden. Legt man das Zeichen des Startspielers fest, z.B: 1, so kann man genauer sagen:

    Die Anzahl an 1en muss größer oder gleich der Anzahl an 2en sein und darf sich maximal um 1 unterscheiden.

Ein möglicher Zustand wäre also z.B: $s = (0,1,2,0,2,2,1,0,1)$, welcher das Spielfeld 
$\begin{array}{c|c|c}
  & x & o \\
  \hline
  & o & o \\
  \hline
x &   & x \\
\end{array}$
darstellt.

### Aktionsraum:
Jede Aktion kann als einzelne natürliche Zahl von 0 bis 8 aufgefasst werden.

Diese Zahlen entsprechen den Feldern im 3x3 Raster wie folgt: $\begin{array}{c|c|c}
 0 & 1 & 2 \\
  \hline
 3 & 4 & 5 \\
  \hline
 6 & 7  & 8 \\
\end{array}$

Der Aktionsraum ist also $A = \{0, 1, 2, 3, 4, 5, 6, 7, 8\}$


Je nach aktuellem Feld sind aber nicht immer alle diese Aktionen möglich, da nicht auf belegte Felder gesetzt werden darf.

# Aufgabe 2

## Käsekästchen

#### Zustandsraum
Platzsparend lässt sich der Zustandsraum darstellen als 3 Arrays.
Der erste Array enthält alle horizontalen Kanten, hat also Dimension 6×5=30,
der zweite Array enthält die vertikalen Kanten, hat also Dimension 7×4=28
und der dritte Array speichert die erzielten Punkte, muss also Dimension 6×4=24 haben.

Die Einträge im ersten und zweiten Array sind 0, wenn die entsprechende Kante noch nicht gezeichnet wurde und 1, wenn sie gezeichnet wurde.
Die Einträge im dritten Array sind 0, wenn auf dem entsprechenden Feld noch kein Punkt erzeugt wurde, 1 wenn der erste Spieler einen Punkt auf diesem Feld erzielt hat und -1 wenn der zweite Spieler einen Punkt auf diesem Feld erzielt hat.

In [2]:
def freshDotsNLines ():
    return ([0]*30, [0]*28, [0]*24)

#### Aktionsraum
Mögliche Aktionen sind das Zeichnen einer horizontalen oder vertikalen Kante, die nicht bereits gezeichnet wurde. Eine Aktion wird dann dargestellt als ein Tupel. Die erste Komponente ist 0 für eine horizontale und 1 für eine vertikale Kante, die zweite Komponente ist der Index in den zugehörigen Array von Kanten. Kanten, die bereits gezeichnet wurden, müssen ausgefiltert werden.

Sonst gibt es keine Einschränkungen der möglichen Aktionen.

In [3]:
def dotsNLinesActions (horizontals, verticals):
    actions = []
    for i in range(30):
        if horizontals[i] != 1:
            actions.append([(0,i)])
    for i in range(28):
        if verticals[i] != 1:
            actions.append([(1,i)])
    return actions

## Vier gewinnt

Wir können einfach die Menge an möglichen Aktionen berechnen indem geprüft wird ob eine Spalte voll ist.

Da die Spielsteine immer spaltenweise gesetzt werden wird das Spielfeld auch Spaltenweise und nicht, wie sonst üblich, zeilenweise gespeichert.

Dies erlaubt die Vereinfachung Listenlängen zu nutzen um zu prüfen wie viele Steine in einer Spalte sind. Außerdem muss man so leere Felder nicht direkt speichern.

Zum einfacheren Prüfen ob jemand gewonnen hat werden dann strings statt Listen verwendet. Dabei werden auch die oben beschriebenen Zahlen 0,1,2 durch die Zeichen `' '`, `'o'`, `'x'` ersetzt.

In [37]:
def get_actions(field):
    """
    inputs:
        field - (list) of 7 strings - maximum length 6, elements are characters 'x' and 'o'
    returns:
        (list) - column indeces of all possible moves
    """
    actions = []
    if not game_ended(field):
        for i, column in enumerate(field):
            if len(column) < 6:
                actions.append(i)
    return actions


def game_ended(field):
    """
    check whether or not someone has won on the current field
    inputs:
        field - (list) of 7 strings - maximum length 6, elements are characters 'x' and 'o'
    returns:
        (bool)
    """
    field = extend_field(field)
    rows = ["".join([field[c][r] for c in range(7)]).strip(" ") for r in range(6)]
    columns = field
    diagonals = get_diagonals(field) + get_diagonals(list(reversed(field)))
    
    for line in rows+columns+diagonals:
        if len(line) < 4:
            continue
        if "xxxx" in line or "oooo" in line:
            return True
    return False
    
    
def get_diagonals(field):
    """
    return strings containing all elements on the y=-x diagonals
    inputs:
        field - (list) of 7 strings - strings must have length 6 and characters ' ','x' or 'o'
    """
    diagonals = []
    for x0, y0 in [(0,3), (0,4), (0,5), (1,5), (2,5), (3,5)]:
        diagonal =  ""
        while True:
            try:
                diagonal += field[x0][y0]
            except IndexError:
                break
            x0 += 1
            y0 -= 1
        diagonals.append(diagonal.strip(" "))
    return diagonals
    

def extend_field(field):
    """
    extend strings in the field to the maximum length of 6 by adding spaces
    returns:
        (list) - new field
    """
    new_field = []
    for column in field:
        new_field.append(column + " "*(6-len(column)))
    return new_field

In `game_ended` müssen auch die Diagonalen geprüft werden. Dafür wird von den im Bild gezeigten Randpunkten aus jeweils jedes Element rechts darunter der Diagonale hinzugefügt, bis das Feld endet.

Durch erneutes Ausführen mit umgekehrter Spaltenreihenfolge werden die anderen Diagonalen getroffen.

Hier ist es wichtig das vollständige Feld zu betrachten und nicht nur die gekürzten Listen. Andernfalls könnten gewinnbringende Diagonalen übersehen werden. Dies wird durch `extend_field` umgesetzt

<img src="vier_gewinnt.JPG" alt="Drawing" style="width: 400px;"/>

zum einfacheren Testen wird noch eine Darstellungs-Funktion implementiert:

In [60]:
def print_field(field):
    """
    print field pretty
    inputs:
        field - (list) of 7 strings - maximum length of strings: 6
    """
    field = extend_field(field)
    print(9*"-")
    for row in range(5,-1,-1):
        print("|{}{}{}{}{}{}{}|".format(*[col[row] for col in field]))
    print(9*"-")

Jetzt haben wir also eine Funktion `get_actions()`, die alle möglichen Aktionen in Abhängigkeit des momentanen Zustands ausgibt und berücksichtigt, wenn ein Spiel schon vorbei ist oder Spalten voll sind. Damit könnte genau wie unten auch eine Methode zum Generieren aller Zustände implementiert werden.

Anhand der folgenden zwei Beispiele wird einige FUnktionalität getestet. Das zweite kann sehr einfach angepasst werden um verschiedene Scenarien für die Gewinnposition zu testen.

In [52]:
field = ["", "o", "xoooxo", "ox", "x", "", "x"]
print_field(field)
print("posiible actions:", get_actions(field))

---------
|  o    |
|  x    |
|  o    |
|  o    |
|  ox   |
| oxox x|
---------
posiible actions: [0, 1, 3, 4, 5, 6]


In [59]:
field = ["xo", "ooox", "xox", "ox", "x", "xo", "xo"] # diagonal
# field = ["xo", "oooo", "xox", "ox", "x", "xx", "xo"] # vertical
# field = ["xo", "ooox", "xxo", "xo", "x", "xo", "xo"] # horizontal
print_field(field)
print("posiible actions:", get_actions(field))

---------
|       |
|       |
| x     |
| ox    |
|ooox oo|
|xoxoxxx|
---------
posiible actions: []


## Tic Tac Toe

Zuerst definieren wir zwei Methoden um einen Zustand besser lesbar ausgeben zu können:

In [3]:
def print_field(field):
    """
    print the given field
    inputs:
        field - (list) - list of 9 ints in (0,1,2)
            0 is an empty field
    """
    print_list = convert_field(field)
    for i in range(3):
        print("-"*13)
        print("| {} | {} | {} |".format( *print_list[3*i:3*i+3] ))
    print("-"*13)

def convert_field(field):
    """
    prepare field for printing by replacing characters
    inputs:
        field - (list) - list of 9 ints in (0,1,2)
            0 is an empty field
    returns:
        (list) - list of "o", "x" and " " representing the field
    """
    print_list = []
    for elem in field:
        if elem == 1:
            print_list.append("o")
        elif elem == 2:
            print_list.append("x")
        else:
            print_list.append(" ")
    return print_list

Dann brauchen wir zwei Funktionen um für ein gegebenes Feld alle möglichen Züge auszugeben. Hat beim gegebenen Feld bereits ein Spieler gewonnen oder ist das Feld voll, so ist keine weitere Aktion möglich. Sonst kann in jedes freie Feld (d.h. Listeneintrag 0) gesetzt werden.

In [4]:
from math import sqrt
def game_ended(field):
    """
    checks whether or not a ttt-game has ended
    inputs:
        field - (list) - list with integer entries 0, 1 and 2
    ouputs:
        (bool) - 
    """
    n = int(sqrt(len(field)))
    rows = [field[n*i:n*i+n] for i in range(n)] #get all rows of the field
    columns = [field[i::n] for i in range(n)]   #get all columns of the field
    diagonals = [field[::n+1], field[n-1:n**2-2:n-1]] #get both diagonals of the field
    
    win_lists = rows + columns + diagonals

    if [1]*n in win_lists or [2]*n in win_lists:
        return True
    return False


def get_ttt_actions(field):
    """
    calculate all possible actions for the given field
    inputs:
        field - (list) - list of 9 ints in (0,1,2)
            0 is an empty field
    returns:
        (list) - list of possible actions as list indices
    """
    actions = []
    if 0 in field and not game_ended(field):
        for i, elem in enumerate(field):
            if elem == 0:
                actions.append(i)
    return actions

Dies können wir nun an einem einfachen Beispiel testen:

In [5]:
field = [0,2,1,0,1,2,2,1,0]
print_field(field)
print("possible actions are:", str(get_ttt_actions(field))[1:-1] )

-------------
|   | x | o |
-------------
|   | o | x |
-------------
| x | o |   |
-------------
possible actions are: 0, 3, 8


Mit dieser Vorbereitung wollen wir nun alle möglichen Zustände generieren.

Dazu gehen wir im Folgenden davon aus, dass der startende Spieler immer Zeichen 1 hat (oben "o").

Durch Zählen der Zahlen 1 und 2 kann immer das Zeichen des nächsten Spielers ermittelt werden. Kommen beide Zahlen gleich oft vor, so ändert sich nichts am Startwert `1` von `active_color`. Kommt das Zeichen `1` öfter vor, so ergibt sich `active_color = 2`.

Dann wird für alle verfügbaren Aktionen der neue Zustand generiert und als Tupel in `states` abgespeichert. Durch Verwendung einer Menge vermeiden wir doppeltes Zählen gleicher Zustände. Ist ein Zustand gespeichert werden rekursiv alle möglichen Folgezustände berechnet und das Ergebnis mit der aktuellen Menge vereinigt (in python `|`).

Die Rekursion endet, wenn bei einem Feld keine Züge mehr mööglich sind.

In [6]:
def get_state_space(field = [0 for _ in range(9)], states = set()):
    """
    return all possible states of a tic-tac-toe game
    inputs: 
        None
    returns:
        (list) of tuples - list of all possible states as tuples of integers (0,1,2)
    """
    active_color = 1
    for elem in field: # calculate color of active player
        if elem == 1:
            active_color += 1
        elif elem == 2:
            active_color -= 1
    for move_index in get_ttt_actions(field): # loop over every possible move
        # if someone has won, this list of moves is empty, therefor ending the recursion
        new_field = [elem for elem in field] # copy the field
        new_field[move_index] = active_color # make the current move
        states.add(tuple(new_field))             # save the current state
        states |= get_state_space(field=new_field, states=states) #add further states to the set
    return states

In [7]:
S = get_state_space()
print(len(S))

5477


Es gibt also 5477 mögliche Zustände. Dies beinhaltet alle Zustände, die durch Rotationen, Spiegelungen oder durch Vertauschen beider Symbole ineinander überführt werden können, aber keine exakten Dopplungen.

Dies sind deutlich weniger als die oben abgeschätzten $3^9 = 19683$ Zustände. Die Spielregeln stellen also eine sehr starke Einschränkung dar.

In [8]:
# print a few example states
for field in list(S)[::2000]:
    print_field(field)
    print("possible actions:", get_ttt_actions(field))
    print()

-------------
| o | x |   |
-------------
| x | x | o |
-------------
| o |   | o |
-------------
possible actions: [2, 7]

-------------
| o | x |   |
-------------
| x | x |   |
-------------
| o | o |   |
-------------
possible actions: [2, 5, 8]

-------------
| o | o | x |
-------------
| x | x | o |
-------------
|   | x | o |
-------------
possible actions: [6]

